### PR TITLE
[Lang] MatrixNdarray refactor part8: Add scalarization for BinaryOpStmt with TensorType-operands

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = https://github.com/taichi-dev/taichi/issues/5873, https://github.com/taichi-dev/taichi/issues/5819

This PR is working "Part ④" in https://github.com/taichi-dev/taichi/issues/5873.

### CODE REVIEW

**Comment by AD1024** [comment]
> Let @AD1024's PR #6077 go first

Merged! Thanks!

File: taichi/transforms/scalarize.cpp [hunk 1]
```diff
@@ -178,6 +177,93 @@ class Scalarize : public IRVisitor {
...
+      tmp = MatrixInitStmt(calc_val0, calc_val1,
+                           calc_val2, calc_val3)
+
+      stmt->replace_all_usages_with(tmp)
+  */
+  void visit(BinaryOpStmt *stmt) override {
+    auto lhs_dtype = stmt->lhs->ret_type;
+    auto rhs_dtype = stmt->rhs->ret_type;
+
+    if (lhs_dtype->is<TensorType>() || rhs_dtype->is<TensorType>()) {
```
Comment (strongoier) [inline taichi/transforms/scalarize.cpp:207]: After https://github.com/taichi-dev/taichi/blob/e8dc076cfcceee10ae93245afde81c1e01b88724/taichi/ir/frontend_ir.cpp#L247-L249 we can assume both lhs and rhs are `TensorType` with same shape, so most logic here is unnecessary now.

File: python/taichi/lang/matrix.py [hunk 2]
```diff
@@ -502,6 +502,8 @@ def __init__(self,
 
     def _element_wise_binary(self, foo, other):
         other = self._broadcast_copy(other)
+        if in_taichi_scope() and impl.current_cfg().real_matrix:
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply python/taichi/lang/matrix.py:505]: Thanks! Haven't realized we no longer go through `_element_wise_binary` with `real_matrix=True`

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**